src/main/java/jenkins/plugins/git/GitToolChooser.java

```diff
@@ -86,16 +87,45 @@ public void addListener(TaskListener listener) {
         boolean useCache = false;
-        String cacheEntry = AbstractGitSCMSource.getCacheEntry(remoteName);
-        File cacheDir = AbstractGitSCMSource.getCacheDir(cacheEntry, false);
-        if (cacheDir != null) {
-            Git git = Git.with(TaskListener.NULL, new EnvVars(EnvVars.masterEnvVars)).in(cacheDir).using("git");
-            GitClient client = git.getClient();
-            if (client.hasGitRepo()) {
-                sizeOfRepo = FileUtils.sizeOfDirectory(cacheDir);
-                sizeOfRepo = (sizeOfRepo/1000); // Conversion from Bytes to Kilo Bytes
-                useCache = true;
+        if (setSizeFromInternalCache(remoteName)) {
+            LOGGER.log(Level.INFO,
+                    "Found cache key for {0} with size {1}",
+                    new Object[]{remoteName, sizeOfRepo});
+            useCache = true;
+            return useCache;
+        }
+        for (String repoUrl : remoteAlternatives(remoteName)) {
+            String cacheEntry = AbstractGitSCMSource.getCacheEntry(repoUrl);
+            File cacheDir = AbstractGitSCMSource.getCacheDir(cacheEntry, false);
+            if (cacheDir != null) {
+                Git git = Git.with(TaskListener.NULL, new EnvVars(EnvVars.masterEnvVars)).in(cacheDir).using("git");
+                GitClient client = git.getClient();
+                if (client.hasGitRepo()) {
+                    long clientRepoSize = FileUtils.sizeOfDirectory(cacheDir) / 1024; // Conversion from Bytes to Kilo Bytes
+                    if (clientRepoSize > sizeOfRepo) {
+                        if (sizeOfRepo > 0) {
+                            LOGGER.log(Level.FINE, "Replacing prior size estimate {0} with new size estimate {1} for remote {2} from cache {3}",
+                                    new Object[]{sizeOfRepo, clientRepoSize, remoteName, cacheDir});
+                        }
+                        sizeOfRepo = clientRepoSize;
+                        assignSizeToInternalCache(remoteName, sizeOfRepo);
+                    }
+                    useCache = true;
+                    if (remoteName.equals(repoUrl)) {
+                        LOGGER.log(Level.FINE, "Remote URL {0} found cache {1} with size {2}",
+                                new Object[]{remoteName, cacheDir, sizeOfRepo});
+                    } else {
+                        LOGGER.log(Level.FINE, "Remote URL {0} found cache {1} with size {2}, alternative URL {3}",
+                                new Object[]{remoteName, cacheDir, sizeOfRepo, repoUrl});
+                    }
+                } else {
+                    // Log the surprise but continue looking for a cache
+                    LOGGER.log(Level.FINE, "Remote URL {0} cache {1} has no git dir", new Object[]{remoteName, cacheDir});
+                }
             }
+        if (!useCache) {
+            LOGGER.log(Level.FINE, "Remote URL {0} cache not found", remoteName);
+        }
         }
```

Comment:
Should only log that cache is not found after checking all variants, not when checking each of the variants
```suggestion
        }
        if (!useCache) {
            LOGGER.log(Level.FINE, "Remote URL {0} cache not found", remoteName);
        }
```